### ISSUE TITLE
Document API end-points, JSON schemas and stability expectations

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/pub-dartlang-dart#2142]
We should write a markdown document that documents APIs we offer, and outline the stability of these APIs.

Notably:
 * Get versions of a package (stable)
 * Get a package (stable)
 * Search for packages with JSON API (might be experimental)

Do we have other APIs that we want to expose to the world?

Note. Maybe we can declare the search API stable, it would certainly be useful to be able to list all packages and things like that. We don't have to make all features of the search API stable, ie. all filters that can be applied (I could imagine those changing over time).

**Comment by isoos** [comment]
We do have a few more, I'll need to go through the code to get all of it.

I'm wondering if this could also be reflected in (a) tests, or documentation for tests, and (b) in the code that defines the routes. Assuming we migrate to use `shelf_router`, would it make sense to annotate the routes with something that reflects their stability?

**Comment by jonasfj** [comment]
> Assuming we migrate to use shelf_router, would it make sense to annotate the routes with something that reflects their stability?

Yes, it would also make sense to write documentation comments that can be extracted into a reference for the API... But this all requires a lot of tooling, let's take on step at the time.

> We do have a few more, I'll need to go through the code to get all of it.

We don't have to document all APIs, ideally we document those that external parties depend on. Or the ones we want to offer people to program against. It's still okay to have internal API end-points for which we make no stability promises.

**Comment by jingtuo** [comment]
> * experimental
Does the search API have more parameters, such as the number of pages per page, filtering packages like dart-io?

For example, if I search for json, I don't want to see dart:io in the results.

**Comment by svenjacobs** [comment]
Is there an API endpoint for fetching the changelog(s) of a package?

**Comment by isoos** [comment]
@svenjacobs: There is none at the moment (unless you count `/packages/<pkg>/changelog`). Curious: what's the use case you are looking for?

**Comment by svenjacobs** [comment]
> Curious: what's the use case you are looking for?

I'm thinking of writing a graphical version of `pub outdated` that also displays all changelogs of outdated library versions starting from _current_ version (+1) to _upgradable_ giving a nice overview of what has changed.

**Comment by sigurdm** [comment]
We have pub.dev/help/api
